frontend/packages/console-dynamic-plugin-sdk/src/utils/fetch/co-fetch.ts

```diff
@@ -0,0 +1,141 @@
+import * as _ from 'lodash';
+import 'whatwg-fetch';
+import { CoFetchText, CoFetchJSON, CoFetch } from '../../api/api-types';
+import { TimeoutError, RetryError } from '../error/http-error';
+import { getCSRFToken, validateStatus, getImpersonateHeaders } from './co-fetch-utils';
+
+const initDefaults = {
+  headers: {},
+  credentials: 'same-origin',
+};
+
+const coFetchInternal = (
+  url: string,
+  options: RequestInit,
+  timeout: number,
+  retry: boolean,
+): Promise<Response> => {
+  const allOptions = _.defaultsDeep({}, initDefaults, options);
+  if (allOptions.method !== 'GET') {
+    allOptions.headers['X-CSRFToken'] = getCSRFToken();
+  }
+
+  // If the URL being requested is absolute (and therefore, not a local request),
+  // remove the authorization header to prevent credentials from leaking.
+  if (url.indexOf('://') >= 0) {
+    delete allOptions.headers.Authorization;
+    delete allOptions.headers['X-CSRFToken'];
+  }
+
+  const fetchPromise = fetch(url, allOptions).then((response) =>
+    validateStatus(response, url, allOptions.method, retry),
+  );
+
+  // return fetch promise directly if timeout <= 0
+  if (timeout < 1) {
+    return fetchPromise;
+  }
+
+  const timeoutPromise: Promise<Response> = new Promise((unused, reject) =>
+    setTimeout(() => reject(new TimeoutError(url, timeout)), timeout),
+  );
+
+  // Initiate both the fetch promise and a timeout promise
+  return Promise.race([fetchPromise, timeoutPromise]);
+};
+
+export const coFetch: CoFetch = async (
+  url: string,
+  options: RequestInit = {},
+  timeout: number = 60000,
+): Promise<Response> => {
```

Comment:
no need for the inline types, its just a duplication because you have `CoFetch` type
```suggestion
export const coFetch: CoFetch = async (url, options = {}, timeout = 60000) => {
```

Reply:
Done.